src/test/java/com/rocketden/main/service/GameManagementServiceTests.java

```diff
@@ -216,7 +242,105 @@ public void submitSolutionSuccess() {
 
         gameService.submitSolution(ROOM_ID, request);
 
-        verify(submitService).submitSolution(any(Game.class), eq(request));
+        verify(submitService).submitSolution(eq(game), eq(request));
+    }
+
+    @Test
+    public void sendAllSolvedSocketUpdate() {
+        Room room = new Room();
+        room.setRoomId(ROOM_ID);
+
+        User user = new User();
+        user.setNickname(NICKNAME);
+        user.setUserId(USER_ID);
+        room.addUser(user);
+
+        User user2 = new User();
+        user2.setNickname(NICKNAME_2);
+        user2.setUserId(USER_ID_2);
+        room.addUser(user2);
+
+        User user3 = new User();
+        user3.setNickname(NICKNAME_3);
+        user3.setUserId(USER_ID_3);
+        room.addUser(user3);
+
+        gameService.createAddGameFromRoom(room);
+        Game game = gameService.getGameFromRoomId(ROOM_ID);
+
+        // Add submissions for the first two users.
+        addSubmissionHelper(game.getPlayers().get(USER_ID), 10);
+        addSubmissionHelper(game.getPlayers().get(USER_ID_2), 10);
+
+        SubmissionRequest request = new SubmissionRequest();
+        request.setLanguage(LANGUAGE);
+        request.setCode(CODE);
+        request.setInitiator(UserMapper.toDto(user3));
+
+        // Mock the return of the submissionDto, and mock update of player.
+        SubmissionDto submissionDto = new SubmissionDto();
+        submissionDto.setNumCorrect(NUM_PROBLEMS);
+        submissionDto.setNumTestCases(NUM_PROBLEMS);
+        Mockito.doAnswer(new Answer<SubmissionDto>() {
+            public SubmissionDto answer(InvocationOnMock invocation) {
+                addSubmissionHelper(game.getPlayers().get(USER_ID_3), 10);
+                return submissionDto;
+            }})
+          .when(submitService).submitSolution(game, request);
+
+        gameService.submitSolution(ROOM_ID, request);
+
+        verify(submitService).submitSolution(eq(game), eq(request));
+
+        // Confirm that socket sent updated GameDto object.
+        verify(socketService).sendSocketUpdate(eq(GameMapper.toDto(game)));
+    }
+
+    @Test
+    public void sendAllSolvedNoSocketUpdate() {
```

Comment:
Possible idea: instead of enforcing the socket update only goes out once, is it possible if we just throw a `GameError.GAME_OVER` error if people try to submit after everyone is done/time is up?

Reply:
I was initially going to implement this, but had another thought -- will this error actually ever exist, or will it just be `GameError.NOT_FOUND`? For example, if the game finishes, wouldn't the game be removed from the `currentGameMap`? Is there ever any intermediary state?

Reply:
That's a good point (I think you commented about it in my other PR which I can implement that clean up stuff) - in that scenario though, would we need this test? Assuming the game is removed when all players finish submitting, then someone submitting again should throw a not found error

Reply:
Not in our current setup -- the situation in which it would be useful is if the game did not end when everybody solves the problem, but rather when the time runs out if players are competing to make the best solution (fastest or least memory). Do you think I should keep this in for that reason, or no?

Reply:
Continued in the other chain